ruby/lib/google/protobuf/descriptor_dsl.rb

```diff
@@ -0,0 +1,440 @@
+#!/usr/bin/ruby
+#
+# Code that implements the DSL for defining proto messages.
+
+require 'google/protobuf/descriptor_pb'
+
+module Google
+  module Protobuf
+    module Internal
+      class Builder
+        def initialize(pool)
+          @pool = pool
+          @default_file = nil  # Constructed lazily
+        end
+
+        def add_file(name, options={}, &block)
+          builder = FileBuilder.new(@pool, name, options)
+          builder.instance_eval(&block)
+          internal_add_file(builder)
+        end
+
+        def add_message(name, &block)
+          internal_default_file.add_message(name, &block)
+        end
+
+        def add_enum(name, &block)
+          internal_default_file.add_enum(name, &block)
+        end
+
+        # ---- Internal methods, not part of the DSL ----
+
+        def build
+          if @default_file
+            internal_add_file(@default_file)
+          end
+        end
+
+        private def internal_add_file(file_builder)
+          proto = file_builder.build
+          serialized = Google::Protobuf::FileDescriptorProto.encode(proto)
+          @pool.add_serialized_file(serialized)
+        end
+
+        private def internal_default_file
+          @default_file ||= FileBuilder.new(@pool, "ruby_default_file.proto")
+        end
+      end
+
+      class FileBuilder
+        def initialize(pool, name, options={})
+          @pool = pool
+          @file_proto = Google::Protobuf::FileDescriptorProto.new(
+            name: name,
+            syntax: options.fetch(:syntax, "proto3")
+          )
+        end
+
+        def add_message(name, &block)
+          builder = MessageBuilder.new(name, self, @file_proto)
+          builder.instance_eval(&block)
+          builder.internal_add_synthetic_oneofs
+        end
+
+        def add_enum(name, &block)
+          EnumBuilder.new(name, @file_proto).instance_eval(&block)
+        end
+
+        # ---- Internal methods, not part of the DSL ----
+
+        # These methods fix up the file descriptor to account for differences
+        # between the DSL and FileDescriptorProto.
+
+        # The DSL can omit a package name; here we infer what the package is if
+        # was not specified.
+        def infer_package(names)
+          # Package is longest common prefix ending in '.', if any.
+          if not names.empty?
+            min, max = names.minmax
+            last_common_dot = nil
+            min.size.times { |i|
+              if min[i] != max[i] then break end
+              if min[i] == ?. then last_common_dot = i end
```

Comment:
let's switch to using strings over character literals, the syntax is confusing to non-ruby folks

Reply:
Done.